File: csharp/LogicAndTrick.WikiCodeParser/Elements/MdCodeElement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Elements
{
    public class MdCodeElement : Element
    {
        private static readonly string[] AllowedLanguages =
        {
            "php", "dos", "bat", "cmd", "css", "cpp", "c", "c++", "cs", "ini", "json", "xml", "html", "angelscript",
            "javascript", "js", "plaintext"
        };

        public MdCodeElement()
        {
            Priority = 10;
        }

        public override bool Matches(Lines lines)
        {
            var value = lines.Value();
            return value.StartsWith("```");
        }

        public override INode Consume(Parser parser, ParseData data, Lines lines, string scope)
        {
            var current = lines.Current();
            var firstLine = lines.Value().Substring(3).TrimEnd();

            string lang = null;
            if (AllowedLanguages.Contains(firstLine, StringComparer.InvariantCultureIgnoreCase))
            {
                lang = firstLine;
                firstLine = "";
            }

            var arr = new List<string> {firstLine};

            var found = false;
            while (lines.Next())
            {
                var value = lines.Value().TrimEnd();
                if (value.EndsWith("```"))
                {
                    var lastLine = value.Substring(0, value.Length - 3);
                    arr.Add(lastLine);
                    found = true;
                    break;
                }
                else
                {
                    arr.Add(value);
                }
            }

            if (!found)
            {
                lines.SetCurrent(current);
                return null;
            }

            // Trim blank lines from the start and end of the array
            for (var i = 0; i < 2; i++)
            {
                while (arr.Count > 0 && arr[0].Trim() == "") arr.RemoveAt(0);
                arr.Reverse();
            }

            // Replace all tabs with 4 spaces
            arr = arr.Select(x => x.Replace("\t", "    ")).ToList();

            // Find the longest common whitespace amongst all lines (ignore blank lines)
            var longestWhitespace = arr.Aggregate(9999, (c, i) =>
            {
                if (i.Trim().Length == 0) return c;
                var wht = i.Length - i.TrimStart().Length;
                return Math.Min(wht, c);
            });

            // Dedent all lines by the longest common whitespace
            arr = arr.Select(a => a.Substring(Math.Min(longestWhitespace, a.Length))).ToList();

            var plain = new UnprocessablePlainTextNode(String.Join("\n", arr));
            var cls = string.IsNullOrWhiteSpace(lang) ? "" : $" class=\"lang-{lang}\"";
            var before = $"<pre{cls}><code>";
            var after = "</code></pre>";
            return new HtmlNode(before, plain, after);
        }
    }
}

File: javascript/src/Elements/MdCodeElement.ts
import { Parser } from '..';
import { Lines } from '../Lines';
import { HtmlNode } from '../Nodes/HtmlNode';
import { INode } from '../Nodes/INode';
import { UnprocessablePlainTextNode } from '../Nodes/UnprocessablePlainTextNode';
import { ParseData } from '../ParseData';
import { Element } from './Element';
import { PreElement } from './PreElement';

export class MdCodeElement extends Element {
    constructor() {
        super();
        this.Priority = 10;
    }

    public Matches(lines: Lines): boolean {
        const value = lines.Value();
        return value.startsWith('```');
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public Consume(parser: Parser, data: ParseData, lines: Lines, _scope: string): INode {
        const current = lines.Current();
        let firstLine = lines.Value().substring(3).trimEnd();

        let lang: string | null = null;
        if (PreElement.AllowedLanguages.includes(firstLine.toLowerCase())) {
            lang = firstLine;
            firstLine = '';
        }

        let arr = [firstLine];

        let found = false;
        while (lines.Next()) {
            const value = lines.Value().trimEnd();
            if (value.endsWith('```')) {
                const lastLine = value.substring(0, value.length - 3);
                arr.push(lastLine);
                found = true;
                break;
            } else {
                arr.push(value);
            }
        }

        if (!found) {
            lines.SetCurrent(current);
            return null;
        }

        // Trim blank lines from the start and end of the array
        for (let i = 0; i < 2; i++) {
            while (arr.length > 0 && arr[0].trim() == '') arr.splice(0, 1);
            arr.reverse();
        }

        // Replace all tabs with 4 spaces
        arr = arr.map(x => x.replace(/\t/g, '    '));

        // Find the longest common whitespace amongst all lines (ignore blank lines)
        const longestWhitespace = arr.reduce((c, i) => {
            if (i.trim().length == 0) return c;
            const wht = i.length - i.trimStart().length;
            return Math.min(wht, c);
        }, 9999);

        // Dedent all lines by the longest common whitespace
        arr = arr.map(a => a.substring(Math.min(longestWhitespace, a.length)));

        const plain = new UnprocessablePlainTextNode(arr.join('\n'));
        const cls = !lang || lang.trim() == '' ? '' : ` class="lang-${lang}"`;
        const before = `<pre${cls}><code>`;
        const after = '</code></pre>';
        return new HtmlNode(before, plain, after);
    }
}